src/libstd/sync/once.rs

```diff
@@ -68,62 +139,183 @@ impl Once {
-        // has run).
-        //
-        // The general scheme here is to gate all future threads once
-        // initialization has completed with a "very negative" count, and to
-        // allow through threads to lock the mutex if they see a non negative
-        // count. For all threads grabbing the mutex, exactly one of them should
-        // be responsible for unlocking the mutex, and this should only be done
-        // once everyone else is done with the mutex.
-        //
-        // This atomicity is achieved by swapping a very negative value into the
-        // shared count when the initialization routine has completed. This will
-        // read the number of threads which will at some point attempt to
-        // acquire the mutex. This count is then squirreled away in a separate
-        // variable, and the last person on the way out of the mutex is then
-        // responsible for destroying the mutex.
-        //
-        // It is crucial that the negative value is swapped in *after* the
-        // initialization routine has completed because otherwise new threads
-        // calling `call_once` will return immediately before the initialization
-        // has completed.
-
-        let prev = self.cnt.fetch_add(1, Ordering::SeqCst);
-        if prev < 0 {
-            // Make sure we never overflow, we'll never have isize::MIN
-            // simultaneous calls to `call_once` to make this value go back to 0
-            self.cnt.store(isize::MIN, Ordering::SeqCst);
+        let mut f = Some(f);
+        self.call_inner(false, &mut |_| f.take().unwrap()());
+    }
+
+    /// Performs the same function as `call_once` except ignores poisoning.
+    ///
+    /// If this `Once` has been poisoned (some initialization panicked) then
+    /// this function will continue to attempt to call initialization functions
+    /// until one of them doesn't panic.
+    ///
+    /// The closure `f` is yielded a boolean indicating whether this `Once` is
+    /// poisoned or not.
+    #[unstable(feature = "once_poison", issue = "31688")]
+    pub fn call_once_force<F>(&'static self, f: F) where F: FnOnce(bool) {
+        // same as above, just with a different parameter to `call_inner`.
+        if self.state.load(Ordering::SeqCst) == COMPLETE {
             return
         }
 
-        // If the count is negative, then someone else finished the job,
-        // otherwise we run the job and record how many people will try to grab
-        // this lock
-        let guard = self.mutex.lock();
-        if self.cnt.load(Ordering::SeqCst) > 0 {
-            f();
-            let prev = self.cnt.swap(isize::MIN, Ordering::SeqCst);
-            self.lock_cnt.store(prev, Ordering::SeqCst);
+        let mut f = Some(f);
+        self.call_inner(true, &mut |p| f.take().unwrap()(p));
+    }
+
+    #[cold]
+    fn call_inner(&'static self,
+                  ignore_poisoning: bool,
+                  mut init: &mut FnMut(bool)) {
```

Comment:
Could you add a comment explaining that we want to function to be virtual to prevent duplication of the cold path, but that there currently isn't an allocation-free way to virtualize an `FnOnce`, which is why this uses `FnMut`?